### PR TITLE
[WebAssembly] Implement an alternative translation for -wasm-enable-sjlj

### PR DESCRIPTION
**Pull request by yamt**
Instead of maintaining per-function-invocation malloc()'ed tables to track which functions each label belongs to, store the equivalent info in jump buffers (jmp_buf) themselves.

Also, use a less emscripten-looking ABI symbols:

    saveSetjmp     -> __wasm_setjmp
    testSetjmp      -> __wasm_setjmp_test
    getTempRet0    -> (removed)
    __wasm_longjmp  -> (no change)

While I want to use this for WASI, it should work for emscripten as well.

An example runtime and a few tests:
https://github.com/yamt/garbage/tree/wasm-sjlj-alt2/wasm/longjmp

wasi-libc version of the runtime:
https://github.com/WebAssembly/wasi-libc/pull/483

emscripten version of the runtime:
https://github.com/emscripten-core/emscripten/pull/21502

Discussion:
https://docs.google.com/document/d/1ZvTPT36K5jjiedF8MCXbEmYjULJjI723aOAks1IdLLg/edit

------

todo:
* ~~remove the option (-mllvm -experimental-wasm-enable-alt-sjlj) by making it unconditionally true~~ done
* ~~rename functions to __wasm_setjmp, __wasm_setjmp_test, __wasm_longjmp~~ done
* ~~update tests~~ done
* ~~update comments~~ done
* ~~update [emscripten PR](https://github.com/emscripten-core/emscripten/pull/21502)~~ done

### CODE REVIEW

**Review by sunfishcode — APPROVED** [review]
This looks good to me! @sbc100 Do you want to coordinate landing this with Emscripten?

**Comment by aheejin** [comment]
Can I take a look before landing this? I'm OOO this afternoon but will be back tomorrow.

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (sbc100) [inline llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp:1018]: This comment should probably be moved to line 1027 and line 1037?

**Review by sbc100 — COMMENTED** [review]
Thanks for working on this!

I'm excited to see this land and become less emscripten-specific. 

I'm hoping that once we get the emscripten side landed we can remove the new option very soon too.

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (yamt) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp]: oops. fixed. thank you.

**Review by sbc100 — COMMENTED** [review]
In terms of getting this landed and tested, I wonder which path we should take:

1.  Land this now, without tests, then update emscripten then come back and flip the default, at which point the existing tests will get updated.
2. Duplicate/update the the existing tests to tests both modes, then delete those changes once we flip the default.

Personally I think I'd be happy with (1) since this is a behind and experimental flag.

What do others think?  @aheejin ?

**Review by aheejin — COMMENTED** [review]
Nice work! Sorry for the late reply. I thought we needed the table because we need to distinguish two different calls from the same function, for example, if `foo` calls `setjmp` and also recursively calls `foo` again, so the same callsite can have two different `setjmpId`s (e.g., https://github.com/emscripten-core/emscripten/blob/main/test/core/test_longjmp3.c). But this fixes that problem by adding `function_invocation_id`, which is simpler and nicer.

I also ran all Emscripten SjLj tests with this implementation (+ your `rt.c`) and all of them seemed to pass. 🎉

---

I left some inline comments. 

+ Given that we don't need `setjmpTableSize` anymore and `setjmpTable` doesn't change, we don't need the whole block here from line 1463 ~ line 1503 doing SSA updates anymore: 
https://github.com/llvm/llvm-project/blob/578e66ac45dfcc5c739f3525bfb82d71282d925c/llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp#L1409-L1449

So you can wrap this part with `(!EnableWasmAltSjLj)`.

**Comment by aheejin** [comment]
> In terms of getting this landed and tested, I wonder which path we should take:
> 
> 1. Land this now, without tests, then update emscripten then come back and flip the default, at which point the existing tests will get updated.
> 2. Duplicate/update the the existing tests to tests both modes, then delete those changes once we flip the default.
> 
> Personally I think I'd be happy with (1) since this is a behind and experimental flag.
> 
> What do others think? @aheejin ?

Come to think of it, should we even introduce this experimental option? Adding `if (NewOption) ... else ...` everywhere makes the code complicated. Can we just do
1. Add library functions in emscripten
2. Replace the current logic in LLVM with new code. (Without `if`~`else`). Tests can be updated with this.
3. Remove old library functions in emscripten.
?

**Comment by sbc100** [comment]
> > In terms of getting this landed and tested, I wonder which path we should take:
> > 
> > 1. Land this now, without tests, then update emscripten then come back and flip the default, at which point the existing tests will get updated.
> > 2. Duplicate/update the the existing tests to tests both modes, then delete those changes once we flip the default.
> > 
> > Personally I think I'd be happy with (1) since this is a behind and experimental flag.
> > What do others think? @aheejin ?
> 
> Come to think of it, should we even introduce this experimental option? Adding `if (NewOption) ... else ...` everywhere makes the code complicated. Can we just do
> 
> 1. Add library functions in emscripten

So you think it should be possible to have both old and new versions supported in emscripten at the same time?    If that works that would be great.  Do you want to send and emscripten PR?  And link to it here?   If we can confirm that all tests pass then I agree it would be great to do this in a single LLVM change.

> 2. Replace the current logic in LLVM with new code. (Without `if`~`else`). Tests can be updated with this.
> 3. Remove old library functions in emscripten.
>    ?

**Comment by aheejin** [comment]
> > > In terms of getting this landed and tested, I wonder which path we should take:
> > > 
> > > 1. Land this now, without tests, then update emscripten then come back and flip the default, at which point the existing tests will get updated.
> > > 2. Duplicate/update the the existing tests to tests both modes, then delete those changes once we flip the default.
> > > 
> > > Personally I think I'd be happy with (1) since this is a behind and experimental flag.
> > > What do others think? @aheejin ?
> > 
> > 
> > Come to think of it, should we even introduce this experimental option? Adding `if (NewOption) ... else ...` everywhere makes the code complicated. Can we just do
> > 
> > 1. Add library functions in emscripten
> 
> So you think it should be possible to have both old and new versions supported in emscripten at the same time? If that works that would be great. Do you want to send and emscripten PR? And link to it here? If we can confirm that all tests pass then I agree it would be great to do this in a single LLVM change.

No, 1 just adds new library functions to emscripten (with different names than the current ones) and at this point they will not be actually used. The reason we add them there first is otherwise 2 (replacing the current scheme with the new one) wouldn't pass in the emscripten CI. As long as we use different library function names I think it can be fine. (I actually want to reuse the name `__wasm_longjmp` (instead of `__wasm_sjlj_longjmp` this PR uses), but maybe we can change that after landing all these.)

@yamt, do you want to submit a PR to emscripten adding new library functions? I can do that too, but given that this is your code, if you want to do it it'd be good. About where to add them... to make the transition smooth, https://github.com/emscripten-core/emscripten/blob/main/system/lib/compiler-rt/emscripten_setjmp.c looks like a good place to start given that our current functions are here, but we can move them later to elsewhere. WDYT? @sbc100

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (yamt) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp:1337]: we need to insert this to make handleLongjmpableCallsForWasmSjLj happy.
we can simplify it if/when we can retire the current logic.

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (yamt) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp:1338]: ok

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (yamt) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp:1330]: at least in some places (eg `handleLongjmpableCallsForWasmSjLj`) it was easier to share a variable.

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (yamt) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp:1021]: i wanted to have a common prefix for this set of api. (`__wasm_sjlj_`)

what's your suggestion for `testSetjmp`?
```
        saveSetjmp     -> __wasm_setjmp?
        testSetjmp     -> ???
        getTempRet0    -> (removed)
        __wasm_longjmp -> keep it
```

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (yamt) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp:1799]: `SetjmpTableInsts` still has a single item, which is used to pass function_invocation_id.

**Comment by yamt** [comment]
> > In terms of getting this landed and tested, I wonder which path we should take:
> > 
> > 1. Land this now, without tests, then update emscripten then come back and flip the default, at which point the existing tests will get updated.
> > 2. Duplicate/update the the existing tests to tests both modes, then delete those changes once we flip the default.
> > 
> > Personally I think I'd be happy with (1) since this is a behind and experimental flag.
> > What do others think? @aheejin ?
> 
> Come to think of it, should we even introduce this experimental option? Adding `if (NewOption) ... else ...` everywhere makes the code complicated. Can we just do
> 
>     1. Add library functions in emscripten
> 
>     2. Replace the current logic in LLVM with new code. (Without `if`~`else`). Tests can be updated with this.
> 
>     3. Remove old library functions in emscripten.
>        ?

i implemented this way (`if-else`) becasue @sbc100 told me it's necessary for emscripten to support both methods for a short period.

**Comment by yamt** [comment]
> @yamt, do you want to submit a PR to emscripten adding new library functions? I can do that too, but given that this is your code, if you want to do it it'd be good. About where to add them... to make the transition smooth, https://github.com/emscripten-core/emscripten/blob/main/system/lib/compiler-rt/emscripten_setjmp.c looks like a good place to start given that our current functions are here, but we can move them later to elsewhere. WDYT? @sbc100

https://github.com/emscripten-core/emscripten/pull/21502

**Comment by yamt** [comment]
> Given that we don't need `setjmpTableSize` anymore and `setjmpTable` doesn't change, we don't need the whole block here from line 1463 ~ line 1503 doing SSA updates anymore:

i get errors like the following if i simply put the SSA update things in `!EnableWasmAltSjLj` block.
i need to investigate.
```
spacetanuki% /Volumes/PortableSSD/llvm/build/bin/clang --sysroot /opt/wasi-sdk-21.0/share/wasi-sysroot -resource-dir /Volumes/PortableSSD/llvm/llvm/lib/clang/17 --target=wasm32-wasi -Os -c -mllvm -wasm-enable-sjlj -mllvm -experimental-wasm-enable-alt-sjlj a.c  
Instruction does not dominate all uses!
  %val = load i32, ptr %val_gep, align 4
  %setjmp.ret = phi i32 [ 0, %entry.split ], [ %val, %setjmp.dispatch ]
in function f
fatal error: error in backend: Broken function found, compilation aborted!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /Volumes/PortableSSD/llvm/build/bin/clang --sysroot /opt/wasi-sdk-21.0/share/wasi-sysroot -resource-dir /Volumes/PortableSSD/llvm/llvm/lib/clang/17 --target=wasm32-wasi -Os -c -mllvm -wasm-enable-sjlj -mllvm -experimental-wasm-enable-alt-sjlj a.c
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module 'a.c'.
4.      Running pass 'Module Verifier' on function '@f'
```

**Comment by yamt** [comment]
> > Given that we don't need `setjmpTableSize` anymore and `setjmpTable` doesn't change, we don't need the whole block here from line 1463 ~ line 1503 doing SSA updates anymore:
> 
> i get errors like the following if i simply put the SSA update things in `!EnableWasmAltSjLj` block. i need to investigate.
> 
> ```
> spacetanuki% /Volumes/PortableSSD/llvm/build/bin/clang --sysroot /opt/wasi-sdk-21.0/share/wasi-sysroot -resource-dir /Volumes/PortableSSD/llvm/llvm/lib/clang/17 --target=wasm32-wasi -Os -c -mllvm -wasm-enable-sjlj -mllvm -experimental-wasm-enable-alt-sjlj a.c  
> Instruction does not dominate all uses!
>   %val = load i32, ptr %val_gep, align 4
>   %setjmp.ret = phi i32 [ 0, %entry.split ], [ %val, %setjmp.dispatch ]
> in function f
> fatal error: error in backend: Broken function found, compilation aborted!
> PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
> Stack dump:
> 0.      Program arguments: /Volumes/PortableSSD/llvm/build/bin/clang --sysroot /opt/wasi-sdk-21.0/share/wasi-sysroot -resource-dir /Volumes/PortableSSD/llvm/llvm/lib/clang/17 --target=wasm32-wasi -Os -c -mllvm -wasm-enable-sjlj -mllvm -experimental-wasm-enable-alt-sjlj a.c
> 1.      <eof> parser at end of file
> 2.      Code generation
> 3.      Running pass 'Function Pass Manager' on module 'a.c'.
> 4.      Running pass 'Module Verifier' on function '@f'
> ```

i think i found the cause. rebuildSSA is still necessary to propagate "val".

**Comment by aheejin** [comment]
Had a chance to chat with @sbc100 offline and we agreed we can probably do this:

> 1. Add library functions in emscripten
> 2. Replace the current logic in LLVM with new code. (Without `if`~`else`). Tests can be updated with this.
> 3. Remove old library functions in emscripten.

The downside of this approach is when you add library functions in emscripten we can't add tests with them. But given that we will follow up with the LLVM side PRs I think it should be fine. Also this removes the need to duplicate tests in LLVM (or omit tests in LLVM when this lands).

**Comment by aheejin** [comment]
> > > Given that we don't need `setjmpTableSize` anymore and `setjmpTable` doesn't change, we don't need the whole block here from line 1463 ~ line 1503 doing SSA updates anymore:
> > 
> > 
> > i get errors like the following if i simply put the SSA update things in `!EnableWasmAltSjLj` block. i need to investigate.
> > ```
> > spacetanuki% /Volumes/PortableSSD/llvm/build/bin/clang --sysroot /opt/wasi-sdk-21.0/share/wasi-sysroot -resource-dir /Volumes/PortableSSD/llvm/llvm/lib/clang/17 --target=wasm32-wasi -Os -c -mllvm -wasm-enable-sjlj -mllvm -experimental-wasm-enable-alt-sjlj a.c  
> > Instruction does not dominate all uses!
> >   %val = load i32, ptr %val_gep, align 4
> >   %setjmp.ret = phi i32 [ 0, %entry.split ], [ %val, %setjmp.dispatch ]
> > in function f
> > fatal error: error in backend: Broken function found, compilation aborted!
> > PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
> > Stack dump:
> > 0.      Program arguments: /Volumes/PortableSSD/llvm/build/bin/clang --sysroot /opt/wasi-sdk-21.0/share/wasi-sysroot -resource-dir /Volumes/PortableSSD/llvm/llvm/lib/clang/17 --target=wasm32-wasi -Os -c -mllvm -wasm-enable-sjlj -mllvm -experimental-wasm-enable-alt-sjlj a.c
> > 1.      <eof> parser at end of file
> > 2.      Code generation
> > 3.      Running pass 'Function Pass Manager' on module 'a.c'.
> > 4.      Running pass 'Module Verifier' on function '@f'
> > ```
> 
> i think i found the cause. rebuildSSA is still necessary to propagate "val".

Oh, I'm sorry, I was incorrect. We should include this part, because SSA rewriting is necessary for other variables (other than `setjmpTable` and `setjmpTableSize`: https://github.com/llvm/llvm-project/blob/578e66ac45dfcc5c739f3525bfb82d71282d925c/llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp#L1439-L1449

What we can exclude is this part: https://github.com/llvm/llvm-project/blob/578e66ac45dfcc5c739f3525bfb82d71282d925c/llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp#L1409-L1437

I'm sorry for the incorrect comment and unnecessary debugging.. 😢

**Comment by yamt** [comment]
@aheejin @sbc100 
let me confirm the plan on this PR.
i can remove the option `-mllvm -experimental-wasm-enable-alt-sjlj` by making it unconditionally true, and update tests, right?

**Comment by sbc100** [comment]
> @aheejin @sbc100 let me confirm the plan on this PR. i can remove the option `-mllvm -experimental-wasm-enable-alt-sjlj` by making it unconditionally true, and update tests, right?

That is my understanding yes.

**Comment by yamt** [comment]
@aheejin @sbc100
i think i resolved most of your feedback in this PR and the emscripten PR.

i tried to run the relevant emscripten tests. but i have not succeeded yet even w/o these changes. (i'm an emscripten newbie.)

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (sbc100) [inline llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp:204]: Can you include the size that passed to alloca here?  

 (looks like its getInt32Ty() so maybe put 4 here?)

Comment (sbc100) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp:204]: Is the same functionInvocationId used even if there are several calls to setjmp inside the same function?

**Review by sbc100 — COMMENTED** [review]
Let me know if you want some help testing alsongside the emscripten side.

Alternatively myself or @aheejin could perform the emscripten tests.

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (aheejin) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp]: The reason I added all setjmp instructions in `SetjmpTableInsts` was to do the SSA update. Now we have a single `setjmpTable`, we don't need to do it and we don't need `SetjmpTableInsts`. (Sorry I missed this comment before)

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (aheejin) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp]: I think we discussed this in the emscripten PR.

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (aheejin) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp:1308]: Now we don't need to support the old and new options simultaneously we don't need to share this. Also Emscripten SjLj can use your functions too, so it doesn't need it either.

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (aheejin) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp]: I think https://github.com/llvm/llvm-project/pull/84137#discussion_r1524006476 answers it too

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (aheejin) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp:204]: Yes. The combination of function invocation ID + callsite determines the unique `setjmp` call, in my understanding.

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (aheejin) [inline llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp:212]: As I said in the emscripten PR, I think we can use your libraries in both Emscripten SjLj and Wasm SjLj. The only reason we have `handleLongjmpableCallsForEmscriptenSjLj` and `handleLongjmpableCallsForEmscriptenSjLj` separately is how we transform code is different and not because the way we use `saveSetjmp` and `testSetjmp` is any different.

So we can remove all uses of the current `saveSetjmp` and `testSetjmp` in this file and fix test accordingly, and move this comment block to `Emscripten setjmp/longjmp handling` part and leave this Wasm side comment as is.

**Review by aheejin — COMMENTED** [review]
If possible, can you not force-push? It's hard to track changes between commits with force-pushes. Thank you!

**Comment by yamt** [comment]
> If possible, can you not force-push? It's hard to track changes between commits with force-pushes. Thank you!

i had to rebase for some unrelated reasons. i will avoid it next time.

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (yamt) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp]: done

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (yamt) [reply llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp]: yes

**Comment by yamt** [comment]
> i tried to run the relevant emscripten tests. but i have not succeeded yet even w/o these changes. (i'm an emscripten newbie.)

they passed with the latest version of this PR and the emscripten PR.

File: llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp
Comment (aheejin) [inline llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp:1273]:
```suggestion
  // invocations. It's just for pointer comparisons. It will never be
  // dereferenced.
```

**Review by aheejin — COMMENTED** [review]
Nice simplification! Thank you for working on this. Also given that we don't add the new option, we can fix the PR title.

Some minor nits about comments:
- You removed https://github.com/llvm/llvm-project/blob/43fc921795bd130a325c013d60f209b5c6128fc7/llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp#L213-L215 in https://github.com/emscripten-core/emscripten/pull/21502, so I think we should fix this explanation.
- There still seem to be mentions of `testSetjmp`, which should be changed to `__wasm_setjmp_test`.

**Comment by yamt** [comment]
@aheejin i updated the rest of comments

**Comment by yamt** [comment]
@aheejin may i rebase this PR as it now has conflicts?

**Comment by aheejin** [comment]
@yamt Given that this PR is not large and we already went through it anyway, nevermind what I said about rebasing, and please use whatever method you are convenient with. But just FYI  you can do `git merge` to avoid rebasing (I'm not asking you to do it here; just in case you aren't aware)

**Comment by yamt** [comment]
> @yamt Given that this PR is not large and we already went through it anyway, nevermind what I said about rebasing, and please use whatever method you are convenient with. But just FYI you can do `git merge` to avoid rebasing (I'm not asking you to do it here; just in case you aren't aware)

ok. i just rebased + force pushed.

i usually avoid merge commits in a PR because i personally feel it's rather complicated to review.
is it a common practice in this repo?

**Review by aheejin — APPROVED** [review]

**Comment by sbc100** [comment]
Is this good to land now?  @yamt do you have commit access or should @aheejin or myself land this?

**Comment by yamt** [comment]
> I'm not sure about the common practice in this repo, but Github in general supports merge workflows much better. Especially in a long-running large PRs, if you force-push, it's hard for the reviewers to know what has changed since their last review, because all the history his lost and they have to look at the whole diff again. And many PRs in this repo tend to be bigger and long-running than other repos in my opinion.
> 
> Also Github supports a nice feature called "New changed since you last reviewed" image and this feature is usable only with the merge workflow. (The screenshot is not mine; I got it from Google) I wonder why the merge workflow is harder to review.
> 
> Also when you merge a PR, you do "squash and merge", so all those merge commits are not gonna end up in the main branch because they are squashed into a single commit, so you don't need to worry about that. (If you do "Create a merge commit" the story is different but that option is not enabled for this repo and I don't think it's used often elsewhere either)

Ok. Maybe it's fine for projects like this, where we squash commits when landing.

**Comment by yamt** [comment]
> Is this good to land now? @yamt do you have commit access or should @aheejin or myself land this?

I have no commit access. Please go ahead.

**Comment by aheejin** [comment]
OK I'll land this. Thank you for working on this!

**Pull request merged**
